### PR TITLE
Add support for booting from External Flash in AP_Bootloader

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by bugobliterator** [comment]
|method     |  H743/CubeOrange IFlash         |            H757 Ext Flash + ChibiOS in ITCM RAM|
|-----------|---------------------------------|------------------------------------------------|
|nop        |     0.0014 usec/call            |                       0.0015 usec/call|
|micros()   |     0.3579 usec/call            |                       0.2983 usec/call|
|millis()   |     0.5384 usec/call            |                       0.5869 usec/call|
|micros64() |     0.3450 usec/call            |                       0.3378 usec/call|
|fadd       |     0.0192 usec/call            |                       0.0122 usec/call|
|fsub       |     0.0186 usec/call            |                       0.0122 usec/call|
|fmul       |     0.0192 usec/call            |                       0.0120 usec/call|
|fdiv /=    |     0.0212 usec/call            |                       0.0162 usec/call|
|fdiv 2/x   |     0.0440 usec/call            |                       0.0452 usec/call|
|dadd       |     0.0260 usec/call            |                       0.0226 usec/call|
|dsub       |     0.0268 usec/call            |                       0.0196 usec/call|
|dmul       |     0.0318 usec/call            |                       0.0256 usec/call|
|ddiv       |     0.0264 usec/call            |                       0.0200 usec/call|
|sinf()     |     0.3270 usec/call            |                       0.3350 usec/call|
|cosf()     |     0.3170 usec/call            |                       0.3370 usec/call|
|tanf()     |     0.5230 usec/call            |                       0.4380 usec/call|
|acosf()    |     0.3880 usec/call            |                       0.3940 usec/call|
|asinf()    |     0.4390 usec/call            |                       0.4630 usec/call|
|atan2f()   |     0.4500 usec/call            |                       0.5040 usec/call|
|sqrtf()    |     0.1010 usec/call            |                       0.1140 usec/call|
|sin()      |     0.3270 usec/call            |                       0.3230 usec/call|
|cos()      |     0.3100 usec/call            |                       0.3300 usec/call|
|tan()      |     0.5280 usec/call            |                       0.4880 usec/call|
|acos()     |     0.3770 usec/call            |                       0.4020 usec/call|
|asin()     |     0.4440 usec/call            |                       0.5300 usec/call|
|atan2()    |     0.4430 usec/call            |                       0.4740 usec/call|
|sqrt()     |     0.1040 usec/call            |                       0.1120 usec/call|
|sq()       |     0.0310 usec/call            |                       0.0590 usec/call|
|powf(v,2)  |     0.4860 usec/call            |                       0.5770 usec/call|
|powf(v,3.1)|     0.5150 usec/call            |                       0.5470 usec/call|
|EKF        |     22.2600 usec/call           |                       22.4720 usec/call|
|iadd8      |     0.0294 usec/call            |                       0.0144 usec/call    |
|isub8      |     0.0292 usec/call            |                       0.0144 usec/call|
|imul8      |     0.0308 usec/call            |                       0.0166 usec/call|
|idiv8      |     0.0386 usec/call            |                       0.0320 usec/call|
|iadd16     |     0.0314 usec/call            |                       0.0174 usec/call|
|isub16     |     0.0316 usec/call            |                       0.0212 usec/call|
|imul16     |     0.0316 usec/call            |                       0.0172 usec/call|
|idiv16     |     0.0436 usec/call            |                       0.0324 usec/call|
|iadd32     |     0.0160 usec/call            |                       0.0088 usec/call|
|isub32     |     0.0164 usec/call            |                       0.0088 usec/call|
|imul32     |     0.0216 usec/call            |                       0.0148 usec/call|
|idiv32     |     0.0336 usec/call            |                       0.0264 usec/call|
|iadd64     |     0.0334 usec/call            |                       0.0210 usec/call|
|isub64     |     0.0332 usec/call            |                       0.0218 usec/call|
|imul64     |     0.0316 usec/call            |                       0.0268 usec/call|
|idiv64     |     0.1680 usec/call            |                       0.1762 usec/call|
|memcpy128  |     0.8662 usec/call            |                       1.0412 usec/call|
|memset128  |     0.7023 usec/call            |                       1.0249 usec/call|
|delay(1)   |     1004.028 usec/call          |                       1006.416 usec/call|
|SEM        |     0.5326 usec/call            |                       0.6622 usec/call|

**Comment by proficnc** [comment]
Have you got a comparison to onboard flash times?

**Comment by bugobliterator** [comment]
> Have you got a comparison to onboard flash times?

you mean time to flash firmware?

File: Tools/AP_Bootloader/bl_protocol.cpp [hunk 1]
```diff
@@ -79,7 +81,7 @@
 // RESET		finalise flash programming, reset chip and starts application
 //
 
-#define BL_PROTOCOL_VERSION 		5		// The revision of the bootloader protocol
+#define BL_PROTOCOL_VERSION 		6		// The revision of the bootloader protocol
```
Comment (tridge) [inline Tools/AP_Bootloader/bl_protocol.cpp:84]: do we need to change the version? I'm concerned about older bl clients (mission planner, qgc etc)

File: Tools/ardupilotwaf/chibios.py [hunk 2]
```diff
@@ -91,10 +91,79 @@ def run(self):
...
+    # run_str="${OBJCOPY} -O binary ${SRC} ${TGT}"
     always_run = True
+    EXTF_MEMORY_START = 0x90000000
+    EXTF_MEMORY_END  = 0x90FFFFFF
+    INTF_MEMORY_START = 0x08000000
+    INTF_MEMORY_END = 0x08FFFFFF
     def keyword(self):
         return "Generating"
+    def run(self):
+        if self.env.HAS_EXTERNAL_FLASH_SECTIONS:
```
Comment (tridge) [inline Tools/ardupilotwaf/chibios.py:103]: should be broken into two functions not a huge if/else

File: Tools/ardupilotwaf/chibios.py [hunk 3]
```diff
@@ -91,10 +91,79 @@ def run(self):
...
+            # get a list of sections
+            cmd = "{} -A -x {}".format(self.env.get_flat('SIZE'), self.inputs[0])
```
Comment (tridge) [inline Tools/ardupilotwaf/chibios.py:105]: this needs a comment explaining what this is doing

File: Tools/ardupilotwaf/chibios.py [hunk 4]
```diff
@@ -219,7 +294,7 @@ def chibios_firmware(self):
     if self.bld.env.HAVE_INTEL_HEX:
         if os.path.exists(bootloader_bin.abspath()):
             hex_target = self.bld.bldnode.find_or_declare('bin/' + link_output.change_ext('.hex').name)
-            hex_task = self.create_task('build_intel_hex', src=[bin_target, bootloader_bin], tgt=hex_target)
+            hex_task = self.create_task('build_intel_hex', src=[bin_target[0], bootloader_bin], tgt=hex_target)
```
Comment (tridge) [inline Tools/ardupilotwaf/chibios.py:297]: better not to generate the extf file when not needed

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/ardupilotwaf/chibios.py (hunk 4 above)
Comment (bugobliterator) [reply Tools/ardupilotwaf/chibios.py:300]: fixed

**Review by tridge — APPROVED** [review]

**Pull request merged**